src/hotspot/cpu/riscv/gc/shared/barrierSetNMethod_riscv.cpp

```diff
@@ -57,40 +60,63 @@ static int entry_barrier_offset(nmethod* nm) {
   return 0;
 }
 
-class NativeNMethodBarrier: public NativeInstruction {
-  address instruction_address() const { return addr_at(0); }
+class NativeNMethodBarrier {
+  address  _instruction_address;
+  int*     _guard_addr;
+  nmethod* _nm;
+
+  address instruction_address() const { return _instruction_address; }
+
+  int *guard_addr() {
+    return _guard_addr;
+  }
 
   int local_guard_offset(nmethod* nm) {
     // It's the last instruction
     return (-entry_barrier_offset(nm)) - 4;
   }
 
-  int *guard_addr(nmethod* nm) {
-    if (nm->is_compiled_by_c2()) {
-      // With c2 compiled code, the guard is out-of-line in a stub
-      // We find it using the RelocIterator.
-      RelocIterator iter(nm);
-      while (iter.next()) {
-        if (iter.type() == relocInfo::entry_guard_type) {
-          entry_guard_Relocation* const reloc = iter.entry_guard_reloc();
-          return reinterpret_cast<int*>(reloc->addr());
+public:
+  NativeNMethodBarrier(nmethod* nm): _nm(nm) {
+    address barrier_address;
+#if INCLUDE_JVMCI
+    if (nm->is_compiled_by_jvmci()) {
+      _instruction_address = nm->code_begin() + nm->frame_complete_offset();
+      _guard_addr = reinterpret_cast<int*>(nm->consts_begin() + nm->jvmci_nmethod_data()->nmethod_entry_patch_offset());
```

Comment:
I see 'nm->consts_begin()' is used here to calculate '_guard_addr' for the JVMCI case on riscv. Do you have more details about the design? Thanks.

Reply:
I forgot to update the riscv version since Graal isn't actually fully working there.  It should look just like the aarch64 code in this regard as the same strategy should work there too.